compiler/rustc_resolve/src/access_levels.rs

```diff
@@ -46,25 +49,48 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
     /// This will also follow `use` chains (see PrivacyVisitor::set_import_binding_access_level).
     fn set_bindings_access_level(&mut self, module_id: LocalDefId) {
         assert!(self.r.module_map.contains_key(&&module_id.to_def_id()));
-        let module_level = self.r.access_levels.get_access_level(module_id);
-        if !module_level.is_some() {
-            return;
-        }
+
         // Set the given binding access level to `AccessLevel::Public` and
         // sets the rest of the `use` chain to `AccessLevel::Exported` until
         // we hit the actual exported item.
         let set_import_binding_access_level =
-            |this: &mut Self, mut binding: &NameBinding<'a>, mut access_level| {
+            |this: &mut Self, mut binding: &NameBinding<'a>, mut parent_id| {
                 while let NameBindingKind::Import { binding: nested_binding, import, .. } =
                     binding.kind
                 {
-                    this.set_access_level(import.id, access_level);
-                    if let ImportKind::Single { additional_ids, .. } = import.kind {
-                        this.set_access_level(additional_ids.0, access_level);
-                        this.set_access_level(additional_ids.1, access_level);
+                    if let Some(import_id) = this.r.opt_local_def_id(import.id) {
+                        let vis = binding.vis.map_id(Into::into);
+                        this.update_effective_vis(
+                            import_id,
+                            vis,
+                            parent_id,
+                            AccessLevel::Exported,
+                            binding.span,
+                        );
+                        if let ImportKind::Single { additional_ids, .. } = import.kind {
+                            if let Some(id) = this.r.opt_local_def_id(additional_ids.0) {
+                                this.update_effective_vis(
+                                    id,
+                                    vis,
+                                    parent_id,
+                                    AccessLevel::Exported,
+                                    binding.span,
+                                );
+                            }
+
+                            if let Some(id) = this.r.opt_local_def_id(additional_ids.1) {
+                                this.update_effective_vis(
+                                    id,
+                                    vis,
+                                    parent_id,
+                                    AccessLevel::Exported,
+                                    binding.span,
+                                );
+                            }
+                        }
+
+                        parent_id = import_id;
```

Comment:
After this assignment it's not a parent id anymore, are you sure the logic here is correct?